### PR TITLE
[SPARK-4697][YARN]System properties should override environment variables

### PR DESCRIPTION
**Pull request by WangTaoTheTonic**
I found some arguments in yarn module take environment variables before system properties while the latter override the former in core module.

### CODE REVIEW

**Comment by tgravescs** [comment]
@WangTaoTheTonic did you have a chance to test this?

**Comment by WangTaoTheTonic** [comment]
I've tested it and results are:
Before this patch:
Setting|name in client mode|name in cluster mode
---------|-------------------------|---------------------------
none|org.apache.spark.examples.SparkPi|org.apache.spark.examples.SparkPi
spark.app.name|spark.app.name|spark.app.name
SPARK_YARN_APP_NAME|SPARK_YARN_APP_NAME|org.apache.spark.examples.SparkPi
spark.app.name and SPARK_YARN_APP_NAME|SPARK_YARN_APP_NAME|spark.app.name

After this patch:
Setting|name in client mode|name in cluster mode
---------|-------------------------|---------------------------
none|org.apache.spark.examples.SparkPi|org.apache.spark.examples.SparkPi
spark.app.name|spark.app.name|spark.app.name
SPARK_YARN_APP_NAME|org.apache.spark.examples.SparkPi|org.apache.spark.examples.SparkPi
spark.app.name and SPARK_YARN_APP_NAME|spark.app.name|spark.app.name

As @vanzin said, the `SPARK_YARN_APP_NAME` becomes useless since `spark.app.name` is mandatory.

Note: In test cases I didn't use `SparkConf.setAppName` in application code.

**Comment by vanzin** [comment]
Actually that's a little bit surprising. SparkPi.scala sets the app name to "Spark Pi", so I'd expect that to show up when the user does not explicitly override it. Also, there seems to be some disprepancies between client and cluster mode in handling the env variable.

**Comment by WangTaoTheTonic** [comment]
@vanzin
Note what I note :-)

>Note: In test cases I didn't use SparkConf.setAppName in application code.

**Comment by vanzin** [comment]
I see. Still, the third line of the "before patch" table is unexpected.

**Comment by tgravescs** [comment]
where were you looking to get the application name?  On the RM or in spark or both? 

https://issues.apache.org/jira/browse/SPARK-3678

**Comment by WangTaoTheTonic** [comment]
@vanzin  Which one do you mean? Client mode or Cluster mode?
@tgravescs I looked the name on RM's UI.
I checked SPARK-3678 and realized that if no `spark.app.name` in configuration file or `--name` in command args, in cluster mode it will use `mainClass`.
But in client mode, cause usually we use `SparkConf.setAppName` in application code, so on RM's UI it will show what we set.

**Comment by vanzin** [comment]
I mean it's unexpected because they're different when they should be the same (in that case, the value of `SPARK_YARN_APP_NAME`).

**Comment by WangTaoTheTonic** [comment]
I am afraid it is not.
In cluster mode, in `SparkSubmitArguments.scala` will be assigned with `mainClass` if `spark.app.name` or `--name` is not specified as it will not read `SPARK_YARN_APP_NAME`.
Then in `SparkSubmit.scala` it will pass `args.name` in format of `spark.app.name` and `--name`  to `org.apache.spark.deploy.yarn.Client`.

 `yarn.Client.scala` transform the args to a ClientArguments object, in which `appName` will only get  its value from `--name`.

So, in the progress, the app name would never get value from the env `SPARK_YARN_APP_NAME`. It is only used in client mode.

**Comment by vanzin** [comment]
I understand all that. I'm saying that's unexpected, in that I'd expect both modes to behave the same. So if there's anything to fix here, that's it.

**Comment by WangTaoTheTonic** [comment]
Oh, I see. But after this patch `SPARK_YARN_APP_NAME` becomes useless. It will make behavior in client and cluster mode to be same.
Note that happens when we don't set app name in SparkConf. Otherwise it is a different issue  described in [SPARK-3678](https://issues.apache.org/jira/browse/SPARK-3678). Perhaps we should file another separate PR to solve that.

**Comment by vanzin** [comment]
>  But after this patch SPARK_YARN_APP_NAME becomes useless

That's why we should not commit this patch.

**Comment by WangTaoTheTonic** [comment]
Or we could just make SPARK_YARN_APP_NAME disappear? As the env variable is not recommended and it will cause different behavior. User can still use `spark.app.name`.

Or we should make SPARK_YARN_APP_NAME and spark.app.name a special case in `YarnClientSchedulerBackend.scala` ?

What do you two think? @tgravescs @andrewor14

**Comment by andrewor14** [comment]
Although in general we should honor Spark properties over environment variables, the app name has been a special case and should remain so for backward compatibility. For this PR, I think the goal is to maintain behavior in the "before" table by making more changes in `YarnClientSchedulerBackend`.

Additionally, it is not intuitive that if you set both `SPARK_YARN_APP_NAME` and `spark.app.name`, the behavior is inconsistent between client mode and cluster mode. I think the app name should be a special case for both deploy modes, but we can fix that in a separate PR.

**Comment by WangTaoTheTonic** [comment]
@andrewor14 @vanzin 
After moving the logic to `SparkSubmitArguments`, I tested with no `SparkConf.setAppName` in application code, here is the results:

Setting | name in client mode | name in cluster mode
--------- | ------------------------- | ---------------------------
none | org.apache.spark.examples.SparkPi | org.apache.spark.examples.SparkPi
spark.app.name | spark.app.name | spark.app.name
SPARK_YARN_APP_NAME | SPARK_YARN_APP_NAME | SPARK_YARN_APP_NAME 
spark.app.name and SPARK_YARN_APP_NAME | spark.app.name | spark.app.name

The names on Driver's UI and RM's UI is consistent.

**Comment by tgravescs** [comment]
So it used to be (0.8 and 0.9) that spark on yarn was ran like this.  We kept this work in the 1.0 release as well for backwards compatibility.

spark-class org.apache.spark.deploy.yarn.Client \
        --jar spark-examples.jar \
        --class org.apache.spark.examples.SparkPi --args yarn-standalone --num-workers 3 --worker-memory 2g --queue unfunded

does moving those to spark submit break the app name for this?   I don't know of anyone still using that but if we break it we should consciously decide that is unsupported.

**Comment by WangTaoTheTonic** [comment]
I am afraid in that case user can only set app name via `spark.app.name` or `--name`, `SPARK_YARN_APP_NAME` will not be used.

**Comment by vanzin** [comment]
Personally, since there have been 3 stable releases already that have the current semantics, and we effectively treat env variables as deprecated, I'd just not change anything, document that env variables are deprecated, and remove them in a future release (1.4 or later).

But if you guys really think this is worth doing, the latest patch looks ok, I guess.

**Comment by andrewor14** [comment]
@vanzin I'm not sure if it's on our roadmap to deprecate the environment variables altogether, since there are many others for other modes that are being actively used. That said I would also prefer to remove them eventually, but until then we should keep the semantics in YARN consistent with other modes (i.e. system properties should override env variables). Unfortunately we can't do anything about the app name but to special case it.

@sryza @tgravescs any thoughts?

**Comment by sryza** [comment]
I'm strongly in favor of deprecating the env vars as well.

Until we do that, keeping the semantics consistent (and special casing if need be) seems reasonable to me.

**Comment by tgravescs** [comment]
I'm in favor of deprecating too but I don't think its good to do it in minor release.  Probably 2.0 would be good time.

Lets just leave the app name special case.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala [hunk 1]
```diff
@@ -149,6 +149,10 @@ private[spark] class SparkSubmitArguments(args: Seq[String], env: Map[String, St
     // Global defaults. These should be keep to minimum to avoid confusing behavior.
     master = Option(master).getOrElse("local[*]")
 
+    // In yarn mode, app name can be set via SPARK_YARN_APP_NAME
+    if (master.contains("yarn")) {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: should be `startsWith`, also in the comments we generally capitalize YARN

Comment (sryza) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: Just to be clear, you mean capitalize YARN in the comment, right?

File: yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClientSchedulerBackend.scala [hunk 2]
```diff
@@ -88,18 +87,19 @@ private[spark] class YarnClientSchedulerBackend(
...
           logWarning(s"NOTE: $sparkProp is deprecated. Use ${deprecatedProps(sparkProp)} instead.")
         }
+      } else if (System.getenv(envVar) != null) {
+        extraArgs += (optionName, System.getenv(envVar))
+        if (deprecatedEnvVars.contains(envVar)) {
+          logWarning(s"NOTE: $envVar is deprecated. Use ${deprecatedEnvVars(envVar)} instead.")
+        }
       }
     }
+    sc.getConf.getOption("spark.app.name").foreach(v => extraArgs += ("--name", v))
```
Comment (andrewor14) [inline yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClientSchedulerBackend.scala]: We need to add a comment here to explain why it's a special case

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala (hunk 1 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: yeah

Comment (sryza) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: Woops didn't see your wrote comments

**Comment by andrewor14** [comment]
Hey @WangTaoTheTonic I think this LGTM. By the way, when I merge this I will file a new JIRA that makes the app name behavior consistent across cluster and client modes and add it to the commit title. I'll also fix the few minor comments I left.

**Pull request closed without merging**